docs/source/protocols.rst

```diff
@@ -0,0 +1,572 @@
+.. _protocol-types:
+
+Protocols and structural subtyping
+==================================
+
+The Python type system supports two ways of deciding whether two objects are
+compatible as types: nominal subtyping and structural subtyping.
+
+*Nominal* subtyping is strictly based on the class hierarchy. If class ``Dog``
+inherits class ``Animal``, it's a subtype of ``Animal``. Instances of ``Dog``
+can be used when ``Animal`` instances are expected. This form of subtyping
+subtyping is what Python's type system predominantly uses: it's easy to
+understand and produces clear and concise error messages, and matches how the
+native :py:func:`isinstance <isinstance>` check works -- based on class
+hierarchy.
+
+*Structural* subtyping is based on the operations that can be performed with an
+object. Class ``Dog`` is a structural subtype of class ``Animal`` if the former
+has all attributes and methods of the latter, and with compatible types.
+
+Structural subtyping can be seen as a static equivalent of duck typing, which is
+well known to Python programmers. See :pep:`544` for the detailed specification
+of protocols and structural subtyping in Python.
+
+.. _predefined_protocols:
+
+Predefined protocols
+********************
+
+The :py:mod:`typing` module defines various protocol classes that correspond
+to common Python protocols, such as :py:class:`Iterable[T] <typing.Iterable>`. If a class
+defines a suitable :py:meth:`__iter__ <object.__iter__>` method, type checkers understand that it
+implements the iterable protocol and is compatible with :py:class:`Iterable[T] <typing.Iterable>`.
+For example, ``IntList`` below is iterable, over ``int`` values:
+
+.. code-block:: python
+
+   from typing import Iterator, Iterable, Optional
+
+   class IntList:
+       def __init__(self, value: int, next: Optional['IntList']) -> None:
+           self.value = value
+           self.next = next
```

Comment:
Feel like it would be better if we used a variable name that didn't shadow a builtin function (`next`)